File: Valid Parentheses/myAnswer.java
class Solution {
    public boolean isValid(String s) {
        Stack<Integer> openStack = new Stack<>();   // 열린 괄호를 저장하는 스택
        String textBracket = "({[)}]";              // 괄호 종류를 저장한 텍스트
        int saveIndex;                              // 가장 최근에 열린 괄호의 index
        int i, temp;

        if ((s.length() % 2) == 1) {
            // 문자열이 홀수라면 짝이 확정적으로 맞을 수 없으므로 바로 false
            return false;
        }

        for (i = 0; i < s.length(); i++) {
            temp = textBracket.indexOf(s.substring(i, i+1));  // 현재 입력된 괄호의 종류 확인

            if (temp < 3) {
                // 여는 괄호였을 경우
                openStack.push(temp); // stack에 저장
            }
            else if (openStack.size() == 0) {
                // 닫는 괄호인데 열린 괄호가 앞에 없을 때 (= stack이 비어있을 때)
                // 짝이 안맞다는 의미이므로
                return false;
            }
            else {
                // 닫는 괄호인데 열린 괄호가 앞에 있을 때
                if ((temp - openStack.pop()) != 3) {
                    // 앞서 온 괄호가 같은 종류인지 비교. 다르다면 틀린 것
                    return false;
                }    
            }
        }

        if (openStack.size() != 0) {
            return false; // 마지막으로 짝을 만나지못하고 남은 열린 괄호가 있는지 확인
        }
        
        return true; // 모든 조건 충족시 
    }
}

File: Valid Parentheses/myAnswer.js
var isValid = function(s) {
    var openStack = [];      // 열린 괄호의 저장
    let textBracket = "({[)}]";   // 괄호
    var saveIndex;          // 가장 최근에 열린 괄호의 index
    var i;

    // 홀수씨는 짝이 맞을리 없으니
    if ((s.length%2) === 1)
        return false;

    for (i = 0; i < s.length; i++) {
        temp = textBracket.indexOf(s[i]);
        
        if (temp < 3) {
            // 여는 괄호였을 경우
            openStack.push(temp); // 열린 괄호의 index를 push
        }
        else {
            // 닫는 괄호였을 경우
            if ((temp - openStack.pop()) !== 3) {
                // 가장 최근에 열린 괄호와 닫는 괄호의 차이가 3이 아닌경우 false 반환
                // (3이 아니다 = 같은 종류가 아니다)
                // (index의 값이 서로 같은 종류일 경우 3의 차이가 나기 때문에..)
                return false;
            }
        }
    }

    if (openStack.length !== 0) {
        // 여는 괄호가 모두 닫히지 않았을 경우
        return false; // 닫힘 괄호와 쌍을 이루지 못했다는 의미이므로 false 반환
    }

    return true; // 다 뚫고 여기까지 왔다면 좋아요~
};
